src/coroutine/cancel.rs

```diff
@@ -0,0 +1,74 @@
+use crate::{ffi, Py, PyAny, PyObject};
+use futures_util::future::poll_fn;
+use futures_util::task::AtomicWaker;
+use std::ptr;
+use std::ptr::NonNull;
+use std::sync::atomic::{AtomicPtr, Ordering};
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+#[derive(Debug, Default)]
+struct Inner {
+    exception: AtomicPtr<ffi::PyObject>,
+    waker: AtomicWaker,
+}
+
+/// Helper used to wait and retrieve exception thrown in [`Coroutine`](super::Coroutine).
+///
+/// Only the last exception thrown can be retrieved.
+#[derive(Debug, Default)]
+pub struct CancelHandle(Arc<Inner>);
+
+impl CancelHandle {
+    /// Create a new `CoroutineCancel`.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Returns whether the associated coroutine has been cancelled.
+    pub fn is_cancelled(&self) -> bool {
+        !self.0.exception.load(Ordering::Relaxed).is_null()
+    }
+
+    /// Poll to retrieve the exception thrown in the associated coroutine.
+    pub fn poll_cancelled(&mut self, cx: &mut Context<'_>) -> Poll<PyObject> {
+        // SAFETY: only valid owned pointer are set in `ThrowCallback::throw`
+        let take = || unsafe {
+            // pointer cannot be null because it is checked the line before,
+            // and the swap is protected by `&mut self`
+            Py::from_non_null(
+                NonNull::new(self.0.exception.swap(ptr::null_mut(), Ordering::Relaxed)).unwrap(),
+            )
+        };
+        if self.is_cancelled() {
+            return Poll::Ready(take());
+        }
+        self.0.waker.register(cx.waker());
+        if self.is_cancelled() {
+            return Poll::Ready(take());
+        }
+        Poll::Pending
+    }
+
+    /// Retrieve the exception thrown in the associated coroutine.
+    pub async fn cancelled(&mut self) -> PyObject {
+        poll_fn(|cx| self.poll_cancelled(cx)).await
+    }
+
+    #[doc(hidden)]
+    pub fn throw_callback(&self) -> ThrowCallback {
+        ThrowCallback(self.0.clone())
+    }
+}
+
+#[doc(hidden)]
+pub struct ThrowCallback(Arc<Inner>);
+
+impl ThrowCallback {
+    pub(super) fn throw(&self, exc: &PyAny) {
+        let ptr = self.0.exception.swap(exc.into_ptr(), Ordering::Relaxed);
```

Comment:
`thow` is not public, so it should not be called from multiple threads (it's only called by the coroutine). Also, I don't really understand your point. This is a swap, you could have as much concurrent swap as you want, each of them would retrieve a valid (or null) pointer. The memory models rules mainly concerns R/W synchronization between multiple memory locations, but there is only one here. And the mutex already has an acquire/release semantic, so you don't need more barrier here in my opinion. (And this part of the code is protected by the GIL)

Reply:
> thow is not public, so it should not be called from multiple threads (it's only called by the coroutine). 

It should not, but there nothing enforcing and/or documenting this and things can change. Especially as concurrency in CPython is up for fundamental changes in the medium term future.

> And the mutex already has an acquire/release semantic, so you don't need more barrier here in my opinion. (And this part of the code is protected by the GIL)

Yes the mutex (and the GIL) imply a stronger memory ordering but they are not used everywhere which makes relying on their properties tricky. Personally, I think the potential performance gain from using the relaxed ordering is not worth it to rely on these complexities.

> Also, I don't really understand your point. This is a swap, you could have as much concurrent swap as you want, each of them would retrieve a valid (or null) pointer.

The point is that the pointer alone is not necessarily all the memory which is accessed drop the drop call. For example if the exception object contains pointers to other objects, their state in memory is not necessarily visible on the thread doing the `drop` under relaxed ordering and hence the drop could encounter invalid values. Acquire-release ordering would ensure that any modifications made to memory reachable via the pointer before swapping it in the atomic variable is also visible to the thread swapping it out.

So long story short, I don't think the concurrency situation here is simple enough so that the relaxed ordering is an easy and/or obvious optimization and most likely is not worth it anyway as it does happen next to stronger memory barriers which will inhibit the reordering potential implied by relaxed in any case. But how exactly is far from cleat, at least to me, and hence going for acquire/release seems to be the reasonable choice here IMHO.

Reply:
To make this code simpler, I will include the swap in the mutex scope in the next MR, so there will be no doubt about synchronization.

Reply:
Actually, I would suggest to use only the mutex for now and do anything else as a separate optimization pass. There should be little chance for contention and then a mutex is just an atomic access anyway. This will be obviously correct and side-step any discussion of memory orderings.

Taking a step backwards. Couldn't we just store the exception as `Py<PyAny>` then to further avoid unsafe code?

Reply:
> The point is that the pointer alone is not necessarily all the memory which is accessed drop the drop call. For example if the exception object contains pointers to other objects, their state in memory is not necessarily visible on the thread doing the drop under relaxed ordering and hence the drop could encounter invalid values.

You're 100% right.

> Taking a step backwards. Couldn't we just store the exception as Py<PyAny> then to further avoid unsafe code?

You're 100% right.

Reply:
:+1: Much simpler now.

Let's look into getting that last bit of performance in follow-up changes.